text/2019-03-04-hibernate-raft.md

```diff
@@ -0,0 +1,121 @@
+
+When the group is in Ordered state, and there is no further writes, the group
+can become Idle state. But leader needs to replicate logs to followers, so it
+should not be Idle until all the active followers have received latest logs
+and it applies to latest logs. Note that only active followers need to be
+considered as inactive followers may be isolated, keep ticking may not make a
+difference most of time. Apply to latest logs can ensure all configuration is
+applied.
+
+Because leader only considers active followers, so when inactive followers
+rejoin the group, leader may not replicate logs to the follower in time. If
+the follower is Chaos, it can use vote message to make leader Chaos too, and
+start replication again. If the follower is Idle it may not receive any logs
+until there is write from client again. So leader needs to find a way to see
+if a follower is active again. So even leader is in Idle state, it should
+broadcast heartbeats to all followers in a long period.
+
+Followers can't know whether there is further write, so it can just consider
+the group Idle when it contains logs from current term and has a leader. And
+then let leader to drive the log replication. And even followers are in Idle
+state, it should also check whether heartbeats arrives in long enough time. If
+not, it means the follower may have been removed from the group or leader is
+down. It should become Chaos and campaign.
+
+### RaftLogGc
+
+Because raft log is done by proposal, so only leader needs to tick. And when
+raft tick is stopped, this tick can be stopped too. When new write proposal
+arrives or a new leader is elected, the tick can be started again.
+
+### SplitRegionCheck
+
+This can be handled using the same policy as `RaftLogGc` except that when
+rocksdb finishes compaction or merge/split is finished, it needs to tick again
+at least one time.
+
+### PdHeartbeat
+
+This is also similar to `RaftLogGc`. However, `PdHeartbeat` must tick when
+there are any read proposals to let pd rebalance correctly.
+
+### CheckMerge
+
+It has been implemented in the lazy way already that it's only scheduled when
+a `PreMerge` is proposed while `CommitMerge` is not applied.
+
+### CheckPeerStaleState
+
+This tick can't be removed as there is no reliable way to remove stale peer.
+
+Besides leader has to broadcast heartbeats to all followers in a long period
+as memtioned above, so this tick can be also used for broadcasting and
+checking heartbeats. To make sure followers check the leader heartbeat
+reliablely, we can add a intermediate state to group state as PreChaos. If a
+follower is Idle, then change it to PreChaos; if it's PreChaos, then change it
+to Chaos. So it takes two ticks for follower to actually start campaign while
+it only takes one tick for leader to broadcast heartbeats. So followers should
+generally stay calm if network and clock is good.
+
+## Drawbacks
+
+When a leader is down, it may take a longer time to recover than before.
```

Comment:
What about transit to `Chaos` when a follower receives a client request?

Reply:
It's part of the design, see L44. But client still needs to wait at least a round of election.